### PR TITLE
[hierarchical-compilation][dynamo] Support invoke_subgraph HOP

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138031
* #137808
* #137965
* #138036
* __->__ #137557
* #137538
* #138424
* #137555

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

### CODE REVIEW

**Comment by anijain2305** [comment]
@zou3519 This is ready for review.

File: test/higher_order_ops/test_invoke_subgraph.py [hunk 1]
```diff
@@ -116,5 +118,95 @@ def fn(x):
...
 
 
+@skipIfTorchDynamo("Not a torch._dynamo test")
+class TestInvokeSubgraphCompile(TestCase):
+    def test_simple(self):
+        def gn(x, y):
+            return (torch.mul(x, y),)
+
+        def fn(x, y):
+            return create_invoke_subgraph_op(gn, (x, y))[0]
```
Comment (zou3519) [inline test/higher_order_ops/test_invoke_subgraph.py:128]: So, is the user API? Or is this just an API we're using for testing?

Comment (zou3519) [reply test/higher_order_ops/test_invoke_subgraph.py:128]: If this is the user API, then I want to bikeshed it more, ask for docs, and also ask for input validation (in the previous PR)

File: test/higher_order_ops/test_invoke_subgraph.py [hunk 2]
```diff
@@ -116,5 +118,95 @@ def fn(x):
...
+    def test_nonlocal_update(self):
+        counter = 2
+
+        def gn(x, y):
+            nonlocal counter
+            return (torch.mul(x, y) * counter,)
+
+        def fn(x, y):
+            nonlocal counter
+            a = create_invoke_subgraph_op(gn, (x, y))[0]
```
Comment (zou3519) [inline test/higher_order_ops/test_invoke_subgraph.py:188]: We allow side effects inside the function traced by invoke_subgraph?

File: test/higher_order_ops/test_invoke_subgraph.py [hunk 3]
```diff
@@ -116,5 +118,95 @@ def fn(x):
...
+
+        x = torch.randn(8, requires_grad=True)
+        y = torch.randn(8, requires_grad=True)
+        ref = fn(x, y)
+
+        counter = 2
+
+        x_clone = x.clone().detach().requires_grad_(True)
+        y_clone = y.clone().detach().requires_grad_(True)
+        res = torch.compile(fn, backend="eager")(x_clone, y_clone)
```
Comment (zou3519) [inline test/higher_order_ops/test_invoke_subgraph.py:200]: can this fullgraph?

File: test/higher_order_ops/test_invoke_subgraph.py (hunk 1 above)
Comment (zou3519) [reply test/higher_order_ops/test_invoke_subgraph.py:128]: It also affects the testing strategy. If this is the frontend, then it should also do something reasonable if the user passes an invalid function to it (like a function that has a graph break in it)

File: torch/_dynamo/variables/higher_order_ops.py [hunk 4]
```diff
@@ -2526,3 +2534,137 @@ def maybe_positional_arg_names(func):
...
+        tx.output.seen_invoke_subgraphs[key] = body_name
+        return body_name
+
+    def call_function(
+        self,
+        tx: "InstructionTranslator",
+        args: "List[VariableTracker]",
+        kwargs: "Dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        # This flattens the kwargs into lifted args
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:2641]: So where's the code that's responsible for doing the mutation + aliasing checks?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 5]
```diff
@@ -2526,3 +2534,137 @@ def maybe_positional_arg_names(func):
             else:
                 result.append(name)
     return result
+
+
+def canonicalize(gmod, root_gmod):
+    # autograd_cache_key is sensitive to the name of the placeholder nodes.
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:2540]: Are we sure it's not sensitive to intermediate node names too? Can we add a test for this?

File: test/higher_order_ops/test_invoke_subgraph.py [hunk 6]
```diff
@@ -116,5 +118,95 @@ def fn(x):
...
+        self.assertEqual(y.grad, y_clone.grad)
+
+        # Check that the Dynamo graph has just one subgraph module
+        self.assertEqual(len(backend.graphs), 1)
+        subgraph_attr_names = set()
+        for node in backend.graphs[0].graph.nodes:
+            if node.op == "get_attr":
+                subgraph_attr_names.add(node.target)
+        self.assertEqual(len(subgraph_attr_names), 1)
+
```
Comment (zou3519) [inline test/higher_order_ops/test_invoke_subgraph.py:178]: Can we do an expecttest on the graphs? (I want to see them for one example, just to make sure things are hooked up correctly).

File: torch/_dynamo/variables/higher_order_ops.py [hunk 7]
```diff
@@ -2526,3 +2534,137 @@ def maybe_positional_arg_names(func):
...
+        # Check if the subgraph from speculate_subgraph (body_gmod) and the fake
+        # inputs have already been seen before. If yes, the subgraph is already
+        # installed in the output graph and we can just access the subgraph
+        # using the saved attr name.
+
+        fake_inputs = [arg.as_proxy().node.meta["example_value"] for arg in fn_args_vt]
+
+        key = hash_graph_and_inputs(tx, body_gmod, fake_inputs)
+
+        if key in tx.output.seen_invoke_subgraphs:
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:2626]: Does it make sense to cache this in the same place as all of the other invoke_subgraph caches? (just to centralize them?)

File: test/higher_order_ops/test_invoke_subgraph.py (hunk 1 above)
Comment (anijain2305) [reply test/higher_order_ops/test_invoke_subgraph.py]: Deleted this function for now. We can introduce a user-facing one later on.

File: test/higher_order_ops/test_invoke_subgraph.py (hunk 2 above)
Comment (anijain2305) [reply test/higher_order_ops/test_invoke_subgraph.py]: Yes, these side effects are same as what supported by speculate_subgraph. This test example generates different graphs for the two calls. So, it checks that Dynamo should not incorrectly de-dupe them.

File: test/higher_order_ops/test_invoke_subgraph.py (hunk 3 above)
Comment (anijain2305) [reply test/higher_order_ops/test_invoke_subgraph.py:200]: Yes, good catch.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 4 above)
Comment (anijain2305) [reply torch/_dynamo/variables/higher_order_ops.py:2641]: Higher up in the stack - https://github.com/pytorch/pytorch/pull/138036

File: torch/_dynamo/variables/higher_order_ops.py (hunk 5 above)
Comment (anijain2305) [reply torch/_dynamo/variables/higher_order_ops.py:2540]: We are. The comment is outdated. Let me change and add a test.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 7 above)
Comment (anijain2305) [reply torch/_dynamo/variables/higher_order_ops.py:2626]: Yes, that indeed looks better. If you don't mind, we can merge this one. And then I can move this in https://github.com/pytorch/pytorch/pull/137965

File: test/higher_order_ops/test_invoke_subgraph.py (hunk 2 above)
Comment (anijain2305) [reply test/higher_order_ops/test_invoke_subgraph.py]: Will add expect test.

**Comment by anijain2305** [comment]
@zou3519 this is ready for another round of review

File: torch/_dynamo/variables/higher_order_ops.py (hunk 7 above)
Comment (zou3519) [reply torch/_dynamo/variables/higher_order_ops.py:2626]: sgtm

File: test/higher_order_ops/test_invoke_subgraph.py [hunk 8]
```diff
@@ -116,5 +118,209 @@ def fn(x):
...
+            return (child,)
+""",
+        )
+
+    def test_nonlocal_update(self):
+        counter = 2
+
+        def gn(x, y):
+            nonlocal counter
+            return (torch.mul(x, y) * counter,)
```
Comment (zou3519) [inline test/higher_order_ops/test_invoke_subgraph.py:208]: what happens if you mutate counter here? Does that error out?

**Review by zou3519 — COMMENTED** [review]
Code looks fine to me. I do want to check on what happens with side effects here (added a question). Also your tests might be failing

**Comment by anijain2305** [comment]
Merged in https://github.com/pytorch/pytorch/pull/137538 because of hop_db test requirements.

**Pull request closed without merging**